### PR TITLE
Kserve v2 - Fix for importlib error

### PR DESCRIPTION
**Pull request by shrinath-suresh**
Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

## Description

[Torchserve KFS Dockerfile](https://github.com/pytorch/serve/blob/master/kubernetes/kserve/Dockerfile.dev) uses python 3.6. 

```
from importlib.metadata import version
```

fails in version 3.6 and works only in python 3.8.

Hence the docker image throws the following exception
```
sudo docker run -it shrinathsuresh/torchserve-kfs:dev
Traceback (most recent call last):
  File "/home/model-server/kserve_wrapper/__main__.py", line 4, in <module>
    from importlib.metadata import version
ModuleNotFoundError: No module named 'importlib.metadata'
```

**Updated the docker file to install python 3.8**

**Steps to test the PR (locally)**

1. Download the docker image - `docker pull shrinathsuresh/torchserve-kfs:dev`
2. Clone the torchserve repo and run the following command to create the mar file
```
torch-model-archiver --model-name mnist --version 1.0 --model-file examples/image_classifier/mnist/mnist.py --serialized-file examples/image_classifier/mnist/mnist_cnn.pt --handler  examples/image_classifier/mnist/mnist_handler.py
```
3. move the mar file to `/mnt/models/model-store`
```
sudo mv mnist.mar /mnt/models/model-store
```
4. Create the following config properties file in `/mnt/models/config/config.properties`
```
inference_address=http://0.0.0.0:8085
management_address=http://0.0.0.0:8085
metrics_address=http://0.0.0.0:8082
enable_envvars_config=true
#grpc_inference_port=7070
#grpc_management_port=7071
install_py_dep_per_model=true
enable_metrics_api=true
service_envelope=kservev2
metrics_format=prometheus
NUM_WORKERS=1
number_of_netty_threads=4
job_queue_size=10
model_store=/mnt/models/model-store
model_snapshot={"name":"startup.cfg","modelCount":1,"models":{"mnist":{"1.0":{"defaultVersion":true,"marName":"mnist.mar","minWorkers":1,"maxWorkers":5,"batchSize":1,"maxBatchDelay":5000,"responseTimeout":120}}}}
```

5. Start torchserve and wait for model to load
```
torchserve --start --ts-config /mnt/models/config/config.properties
```
6. Run the following command to start kserve using torchserve-kfs dev docker image
```
sudo docker run -v /mnt:/mnt --network host -it shrinathsuresh/torchserve-kfs:dev python /home/model-server/kserve_wrapper/__main__.py
```
7. create a sample test file `mnist_v2_bytes.json`
```
{
    "inputs": [
        {
            "data": "iVBORw0KGgoAAAANSUhEUgAAABwAAAAcCAAAAABXZoBIAAAA10lEQVR4nGNgGFhgy6xVdrCszBaLFN/mr28+/QOCr69DMCSnA8WvHti0acu/fx/10OS0X/975CDDw8DA1PDn/1pBVEmLf3+zocy2X/+8USXt/82Ds+/+m4sqeehfOpw97d9VFDmlO++t4JwQNMm6f6sZcEpee2+DR/I4A05J7tt4JJP+IUsu+ncRp6TxO9RAQJY0XvrvMAuypNNHuCTz8n+PzVEcy3DtqgiY1ptx6t8/ewY0yX9ntoDA63//Xs3hQpMMPPsPAv68qmDAAFKXwHIzMzCl6AoAxXp0QujtP+8AAAAASUVORK5CYII=",
            "datatype": "BYTES",
            "name": "312a4eb0-0ca7-4803-a101-a6d2c18486fe",
            "shape": -1
        }
    ]
}
```

8. Run the following curl command for prediction

```
curl -v -H "ContentType: application/json" http://localhost:8080/v2/models/mnist/infer -d @./mnist_v2_bytes.json
```

Sample response

```
{"id": "84c5dc50-a9b8-4f5b-90a7-e94517b795e3", "model_name": "mnist", "model_version": "1.0", "outputs": [{"name": "predict", "shape": [], "datatype": "INT64", "data": [0]}]}
```

Attaching ts_log.log in the logs section

**Steps to test in cluster environment**

Use the following script for cluster creation 

https://github.com/kserve/kserve/blob/master/hack/quick_install.sh 

1. Run the following command to update the torchserve-kfs image

```
kubectl edit cm inferenceservice-config -n kserve -oyaml
```

The pytorch section needs to be updated as follows
```
"pytorch": {
          "image": "shrinathsuresh/torchserve-kfs",
          "defaultImageVersion": "dev",
          "defaultGpuImageVersion": "dev-gpu",
          "supportedFrameworks": [
            "pytorch"
          ],
          "multiModelServer": false
        },

```

2. clone the kserve repository and move to `docs/samples/v1beta1/torchserve/v2`

```
git clone https://github.com/kserve/kserve.git
cd docs/samples/v1beta1/torchserve/v2
```

3. Run the following command to create inference service and wait for the model to get loaded

```
kubectl apply -f torchserve.yaml
```

Model loading can be verified in the logs using the following command  
```
kubectl get logs torchserve-v2-predictor-default-00001-deployment-65c6f57674qjgh
```

4. Once the model is loaded, run the following port forwarding command

```
kubectl port-forward <name> -n <namespace> 8080:80
```

5. Run the curl command for prediction

```
curl -v -H "Host: torchserve-v2.kserve-test.example.com" http://localhost:8080/v2/models/mnist/infer -d @./tensor_conv/mnist_v2.json
```

6. Sample response
```
{"id": "d3b15cad-50a2-4eaf-80ce-8b0a428bd298", "model_name": "mnist", "model_version": "1.0", "outputs": [{"name": "predict", "shape": [1], "datatype": "INT64", "data": [1]}]}
```

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the tests [UT/IT] that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced. 
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- UT/IT execution results

- Logs

Torchserve logs - 
[ts_log.log](https://github.com/pytorch/serve/files/8035287/ts_log.log)

## Checklist:

- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] New and existing unit tests pass locally with these changes?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

**Review by HamidShojanazeri — APPROVED** [review]
Thanks @shrinath-suresh, LGTM.

File: kubernetes/kserve/Dockerfile.dev [hunk 1]
```diff
@@ -28,21 +28,20 @@ RUN --mount=type=cache,id=apt-dev,target=/var/cache/apt \
   dpkg-dev \
   g++ \
   git \
-  python3-dev \
+  python3.8-dev \
```
Comment (lxning) [inline kubernetes/kserve/Dockerfile.dev:31]: Why specific version python3.8-dev is needed?

File: kubernetes/kserve/Dockerfile.dev [hunk 2]
```diff
@@ -28,21 +28,20 @@ RUN --mount=type=cache,id=apt-dev,target=/var/cache/apt \
...
   vim \
   && rm -rf /var/lib/apt/lists/* \
   && cd /tmp \
-  && curl -O https://bootstrap.pypa.io/get-pip.py \
-  && python3 get-pip.py
 
-RUN update-alternatives --install /usr/bin/python python /usr/bin/python3 1 \
-  && update-alternatives --install /usr/local/bin/pip pip /usr/local/bin/pip3 1
+RUN update-alternatives --remove python /usr/bin/python \
+  && update-alternatives --install /usr/bin/python python /usr/bin/python3.8 1
```
Comment (lxning) [inline kubernetes/kserve/Dockerfile.dev:41]: Same here. Why specific version python3.8 is needed?

Comment (shrinath-suresh) [reply kubernetes/kserve/Dockerfile.dev:41]: The base image `ubuntu-18.04` comes with python 3.6. There were two reasons why we wanted to move into new version

We are getting the following error while building the image using python 3.6
```
Traceback (most recent call last):
  File "/home/model-server/kserve_wrapper/__main__.py", line 4, in <module>
    from importlib.metadata import version
ModuleNotFoundError: No module named 'importlib.metadata'
```

Had a discussion with @chauhang and changed it to 3.8 , as 3.6 is bit old and all our kserve testing were done using python 3.8.

**Pull request merged**